src/Interpreters/TLSLog.cpp

```diff
@@ -0,0 +1,218 @@
+}
+
+void fillColumnArray(const Strings & data, IColumn & column)
+{
+    auto & array = typeid_cast<ColumnArray &>(column);
+    size_t size = 0;
+    auto & data_col = array.getData();
+    for (const auto & name : data)
+    {
+        data_col.insertData(name.data(), name.size());
+        ++size;
+    }
+    auto & offsets = array.getOffsets();
+    offsets.push_back(offsets.back() + size);
+}
+
+}
+
+namespace DB
+{
+
+TLSLogElement::TLSLogElement(const String &failure_reason_, uint64_t failure_count_)
+    : type(failure_reason_.empty() ? TLSLogElementType::SESSION_SUCCESS : TLSLogElementType::SESSION_FAILURE), failure_reason(failure_reason_), failure_count(failure_count_)
+{
+    std::tie(event_time, event_time_microseconds) = eventTime();
+}
+
+#if USE_SSL
+void TLSLog::onInvalidCertificate(const void *, Poco::Net::VerificationErrorArgs & errorCert)
+{
+    // Don't put every certificate failure into the log to save disk space and protect from misbehaving clients.
+    // Instead, produce a log per cert serial number once in `TLSLogErrorFlushIntervalSeconds` seconds.
+    std::unique_lock lock{tls_failure_stats_mutex};
+    const String &serial_number = errorCert.certificate().serialNumber();
+    if (!tls_failure_stats.contains(serial_number))
+        tls_failure_stats[serial_number] = std::make_unique<CertStats>(0, std::make_unique<Poco::Net::X509Certificate>(errorCert.certificate()));
+
+    auto &stats = tls_failure_stats[serial_number];
+    ++stats->count;
+    stats->last_error_message = errorCert.errorMessage();
+}
+
+void TLSLog::logTLSConnection(const Poco::Net::X509Certificate &tls_certificate, const String &user)
+{
+    add(TLSLogElement(tls_certificate, user));
+}
+
+void TLSLog::tlsFailureFlushRunner()
+{
+    while (!tls_failure_stop_flush)
+    {
+        try
+        {
+            auto next_flush = std::chrono::system_clock::now() + std::chrono::seconds(TLSLogErrorFlushIntervalSeconds);
+            {
+                std::unique_lock lock{tls_failure_stats_mutex};
+                for (auto &cert_stat : tls_failure_stats)
+                {
+                    auto &stat = cert_stat.second;
+                    add(TLSLogElement(*stat->certificate, "", stat->last_error_message, stat->count));
+                }
+                tls_failure_stats.clear();
+            }
+            std::this_thread::sleep_until(next_flush);
+        }
+        catch (...)
+        {
+            tryLogCurrentException(__PRETTY_FUNCTION__);
+        }
+    }
+}
+#endif
+
+TLSLog::TLSLog(ContextPtr context_, const SystemLogSettings & settings) : SystemLog<TLSLogElement>(context_, settings)
+{
+#if USE_SSL
+    Poco::Net::SSLManager::instance().ServerVerificationError += Poco::Delegate<TLSLog, Poco::Net::VerificationErrorArgs>(this, &TLSLog::onInvalidCertificate);
+    tls_failure_flush_thread = std::make_unique<ThreadFromGlobalPool>([this] { tlsFailureFlushRunner(); });
```

Comment:
You could avoid making the additional thread there if you flush `tls_failure_stats` as part of `SystemLogBase::flush`.

Reply:
as a result there is no need of `TLSLogErrorFlushIntervalSeconds`. The period of system table flash is the period of caching the errors counts.